go/vt/vttablet/tabletmanager/rpc_external_reparent.go

```diff
@@ -105,15 +98,21 @@ func (agent *ActionAgent) TabletExternallyReparented(ctx context.Context, extern
 	}()
 	event.DispatchUpdate(ev, "starting external from tablet (fast)")
 
-	// Execute state change to master by force-updating only the local copy of the
-	// tablet record. The actual record in topo will be updated later.
-	log.Infof("fastTabletExternallyReparented: executing change callback for state change to MASTER")
-	newTablet := proto.Clone(tablet).(*topodatapb.Tablet)
-	newTablet.Type = topodatapb.TabletType_MASTER
+	// only execute state change if we are not already the correct tablet type
+	if tablet.Type != topodatapb.TabletType_MASTER {
+		// Execute state change to master by force-updating only the local copy of the
+		// tablet record. The actual record in topo will be updated later.
+		log.Infof("fastTabletExternallyReparented: executing change callback for state change to MASTER")
+		newTablet := proto.Clone(tablet).(*topodatapb.Tablet)
+		newTablet.Type = topodatapb.TabletType_MASTER
+
+		// This is where updateState will block for gracePeriod, while it gives
+		// vtgate a chance to stop sending replica queries.
+		agent.updateState(ctx, newTablet, "fastTabletExternallyReparented")
+	}
 
-	// This is where updateState will block for gracePeriod, while it gives
-	// vtgate a chance to stop sending replica queries.
-	agent.updateState(ctx, newTablet, "fastTabletExternallyReparented")
+	// always run finalize even if the master tablet has the correct state
```

Comment:
This is starting to be a significant departure from the original design of TER - in particular the split into a fast stage with the short-circuit based on the shard record and a slow "finalize" stage that only runs when something has actually changed.

By recommending TER as a fix for cases when PlannedReparentShard fails, we're also mixing code paths that were never intended to be mixed. TER was supposed to only be used when Vitess-native reparents are disabled.

I'm starting to think what we really need is to make PlannedReparentShard itself more robust. If PRS fails, the solution should be to run PRS again pointing at the same master. It's been a longstanding problem with Vitess reparents that there are no commands to recover from partial operations. I don't think co-opting TER for a purpose outside its intended scope is the right answer.

Reply:
Is this a good summary of the original design?
https://vitess.io/docs/user-guides/reparenting/#external-reparenting
Code in master as of right now doesn't actually perform part2 of step 5 (change tablet type to spare).

Reply:
> It's been a longstanding problem with Vitess reparents that there are no commands to recover from partial operations. 

This is a bit out of scope for this PR, but 👍 I agree completely. Partial Vitess reparents have been tricky for us to manually recover from in the past, and it'd be great to make it safe to retry without having to manually inspect the cluster.

Reply:
> TER was supposed to only be used when Vitess-native reparents are disabled.

This is really interesting history that I didn't know!

Reply:
part2 of step 5 (changing the tablet type of the unresponsive old master to `spare`) seems like it would actually solve a lot of problems we've seen when a master is unavailable for a bit, a TER happens, and then the master comes back online. I would prefer the `spare` (or other non-serving tablet type) to setting it to `replica`.